x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPSeriesWeightedSum.java

```diff
@@ -0,0 +1,176 @@
+        this.field = field;
+        this.p = p;
+    }
+
+    private MvPSeriesWeightedSum(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isType(field, t -> t.isNumeric() && isRepresentable(t), sourceText(), FIRST, "numeric");
+
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        resolution = TypeResolutions.isType(p, dt -> dt == DOUBLE, sourceText(), SECOND, "double")
+            .and(TypeResolutions.isFoldable(p, sourceText(), SECOND));
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return resolution;
+    }
+
+    @Override
+    public boolean foldable() {
+        return field.foldable() && p.foldable();
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        return switch (PlannerUtils.toElementType(field.dataType())) {
+            case DOUBLE -> new MvPSeriesWeightedSumDoubleEvaluator.Factory(
+                source(),
+                toEvaluator.apply(field),
+                ctx -> new CompensatedSum(),
+                (Double) p.fold()
+            );
+            case NULL -> EvalOperator.CONSTANT_NULL_FACTORY;
+
+            default -> throw EsqlIllegalArgumentException.illegalDataType(field.dataType());
+        };
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new MvPSeriesWeightedSum(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MvPSeriesWeightedSum::new, field, p);
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Evaluator(extraName = "Double")
+    static void process(
+        DoubleBlock.Builder builder,
+        int position,
+        DoubleBlock block,
+        @Fixed(includeInToString = false, build = true) CompensatedSum sum,
+        @Fixed double p
+    ) {
+        sum.reset(0, 0);
+        int start = block.getFirstValueIndex(position);
+        int end = block.getValueCount(position) + start;
+
+        for (int i = start; i < end; i++) {
+            double current_score = block.getDouble(i) / Math.pow(i - start + 1, p);
+            sum.add(current_score);
+        }
+        builder.appendDouble(sum.value());
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        Source.EMPTY.writeTo(out);
```

Comment:
Could you write `source()` instead. Writing `EMPTY` was an old pattern that we had a long while ago because our writing of source was really really inefficient. It's much better now.

Reply:
Hey Nik! I tried using `source()`, but the `MvPSeriesWeightedSumSerializationTests` test failed. I don't know why. Could you help me with that?

The error is something like:

```
Expected :sameInstance(<@-1:-1>)
Actual   :<bout the sq@8:27>
```

Reply:
There's an override in there to make it assume you are always sending an empty source:
```
    @Override
    protected boolean alwaysEmptySource() {
        return true;
    }
```

Zap that and it should go.